clients/src/main/java/org/apache/kafka/clients/producer/internals/BuiltInPartitioner.java

```diff
@@ -321,13 +322,20 @@ public int partition() {
         }
     }
 
-    /*
+    /**
      * Default hashing function to choose a partition from the serialized key bytes
      */
     public static int partitionForKey(final byte[] serializedKey, final int numPartitions) {
         return Utils.toPositive(Utils.murmur2(serializedKey)) % numPartitions;
     }
 
+    /**
+     * Default hashing function to choose a partition from the serialized key bytes
+     */
+    public static int partitionForKey(final ByteBuffer serializedKey, final int numPartitions) {
+        return Utils.toPositive(Utils.murmur2(serializedKey)) % numPartitions;
```

Comment:
murmur2() will change the position of the bytebuffer and (correct me if I am wrong) we want to read it again to actually get the key (using remaining()). Wouldn't we end up reading the wrong key since position has already been changed by this method?

Reply:
`Utils#murmur2(ByteBuffer)` does not change the offsets of the input `ByteBuffer` because `Utils#murmur2(ByteBuffer)` only uses the `ByteBuffer#get(int)` and `ByteBuffer#getInt(int)` methods to read data from the `ByteBuffer`, and these methods do not change the offsets of the `ByteBuffer`.

Reply:
Got it.